mongo/integration/unified/testrunner_operation.go

```diff
@@ -111,11 +137,99 @@ func executeTestRunnerOperation(ctx context.Context, operation *operation) error
 		coll := lookupString(args, "collectionName")
 		index := lookupString(args, "indexName")
 		return verifyIndexExists(ctx, db, coll, index, false)
+	case "loop":
+		var unmarshaledArgs loopArgs
+		if err := bson.Unmarshal(args, &unmarshaledArgs); err != nil {
+			return fmt.Errorf("error unmarshalling arguments to loopArgs: %v", err)
+		}
+		return executeLoop(ctx, &unmarshaledArgs, loopDone)
 	default:
 		return fmt.Errorf("unrecognized testRunner operation %q", operation.Name)
 	}
 }
 
+func executeLoop(ctx context.Context, args *loopArgs, loopDone <-chan struct{}) error {
+	// setup entities
+	entityMap := entities(ctx)
+	if args.errorsStored() {
+		if err := entityMap.addBSONArrayEntity(args.ErrorsEntityID); err != nil {
+			return err
+		}
+	}
+	if args.failuresStored() {
+		if err := entityMap.addBSONArrayEntity(args.FailuresEntityID); err != nil {
+			return err
+		}
+	}
+	if args.successesStored() {
+		if err := entityMap.addSuccessesEntity(args.SuccessesEntityID); err != nil {
+			return err
+		}
+	}
+	if args.iterationsStored() {
+		if err := entityMap.addIterationsEntity(args.IterationsEntityID); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-loopDone:
+			return nil
+		default:
+			if args.iterationsStored() {
+				if err := entityMap.incrementIterations(args.IterationsEntityID); err != nil {
+					return err
+				}
+			}
+			var loopErr error
+			for i, operation := range args.Operations {
+				if operation.Name == "loop" {
+					return fmt.Errorf("loop sub-operations should not include loop")
+				}
+				loopErr = operation.execute(ctx, loopDone)
+
+				// if the operation errors, stop this loop
+				if loopErr != nil {
+					// If StoreFailures or StoreErrors is set, continue looping, otherwise break
+					if !args.errorsStored() && !args.failuresStored() {
+						return fmt.Errorf("error running loop operation %v : %v", i, loopErr)
```

Comment:
If an error occurs that breaks the loop when the `loopDone` channel is written too, is there a risk the sender could block?

Edit: Nevermind – I see the note about the panic in `EndLoop`. That seems reasonable.